File: backend/app/schemas/explore.py
"""Schemas for the Explore feature (platform-wide album/group/stats browsing)."""

from datetime import datetime

from pydantic import BaseModel


# ==================== ALBUMS ====================


class ExploreAlbumItem(BaseModel):
    id: int
    spotify_album_id: str | None
    title: str
    artist: str
    artist_url: str | None
    cover_url: str | None
    release_date: str | None
    avg_rating: float | None
    review_count: int
    nomination_count: int
    weighted_score: float | None


class ExploreAlbumsPage(BaseModel):
    items: list[ExploreAlbumItem]
    next_offset: int | None = None


# ==================== GROUPS ====================


class ExploreGroupItem(BaseModel):
    id: int
    name: str
    is_public: bool
    is_global: bool
    member_count: int
    created_at: datetime


class ExploreGroupsPage(BaseModel):
    items: list[ExploreGroupItem]
    next_offset: int | None = None


# ==================== STATS ====================


class ArtistNominationItem(BaseModel):
    artist: str
    artist_url: str | None
    nomination_count: int
    unique_albums: int


class SiteStatsResponse(BaseModel):
    total_albums_nominated: int
    total_reviews: int
    total_active_groups: int
    total_active_members: int
    top_rated_albums: list[ExploreAlbumItem]
    bottom_rated_albums: list[ExploreAlbumItem]
    most_nominated_artists: list[ArtistNominationItem]
    most_nominated_albums: list[ExploreAlbumItem]

File: frontend/src/types/explore.ts
export interface ExploreAlbumItem {
  id: number
  spotify_album_id: string | null
  title: string
  artist: string
  artist_url: string | null
  cover_url: string | null
  release_date: string | null
  avg_rating: number | null
  review_count: number
  nomination_count: number
  weighted_score: number | null
}

export interface ExploreAlbumsPage {
  items: ExploreAlbumItem[]
  next_offset: number | null
}

export interface ExploreAlbumsParams {
  offset?: number
  limit?: number
  min_reviews?: number | null
  sort_by?: string
  q?: string | null
}

export interface ExploreGroupsParams {
  offset?: number
  limit?: number
  q?: string | null
  group_type?: 'all' | 'human' | 'bot'
}

export interface ExploreGroupItem {
  id: number
  name: string
  is_public: boolean
  is_global: boolean
  member_count: number
  created_at: string
}

export interface ExploreGroupsPage {
  items: ExploreGroupItem[]
  next_offset: number | null
}

export interface ExploreUserItem {
  id: number
  username: string
  member_since: string
  review_count: number
  total_groups: number
}

export interface ExploreUsersParams {
  offset?: number
  limit?: number
  q?: string | null
}

export interface ExploreUsersPage {
  items: ExploreUserItem[]
  next_offset: number | null
}

export interface ArtistNominationItem {
  artist: string
  artist_url: string | null
  nomination_count: number
  unique_albums: number
}

export interface SiteStatsResponse {
  total_albums_nominated: number
  total_reviews: number
  total_active_groups: number
  total_active_members: number
  top_rated_albums: ExploreAlbumItem[]
  bottom_rated_albums: ExploreAlbumItem[]
  most_nominated_artists: ArtistNominationItem[]
  most_nominated_albums: ExploreAlbumItem[]
}
